File: problems_list/0039.组合总和/0039-组合总和.cpp
class Solution {
public:
    vector<vector<int>> ans;
    vector<int> rec;

    void dfs(vector<int>& cands, int cur, int target){
        if(target==0){
            ans.push_back(rec);
            return;
        }

        for(int i=cur ;i<(int)cands.size(); i++){
            if(target<cands[i]) break;
            rec.push_back(cands[i]);
            dfs(cands, i, target-cands[i]);
            rec.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        dfs(candidates, 0, target);
        return ans;
    }
};

File: problems_list/0039.组合总和/0039-组合总和.js
var combinationSum = function(candidates, target) {
    const ans = [];
    const dfs = (target, combine, idx) => {
        if (idx === candidates.length) {
            return;
        }
        if (target === 0) {
            ans.push(combine);
            return;
        }
        // 直接跳过
        dfs(target, combine, idx + 1);
        // 选择当前数
        if (target - candidates[idx] >= 0) {
            dfs(target - candidates[idx], [...combine, candidates[idx]], idx);
        }
    }

    dfs(target, [], 0);
    return ans;
};